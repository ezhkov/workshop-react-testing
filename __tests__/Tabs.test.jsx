import React from 'react';
import { mount } from 'enzyme';
import Cookies from 'js-cookie';
import App from '../src/components/App';

jest.mock('js-cookie');

const tabSelector = 'li[data-test="tab"]';
const tabListSelector = 'ul[data-test="tabList"]';
const tabPanelSelector = 'div[data-test="tabPanel"]';
const tabContentSelector = 'div[data-test="tabContent"]';
const tabRemoveSelector = '[data-test="removeTab"]';
const tabAddSelector = '[data-test="addTab"]';

const appSelector = wrapper => ({
  getTabs: () => wrapper.find(tabSelector),
  getTabsList: () => wrapper.find(tabListSelector),
  getTabsContents: () => wrapper.find(tabPanelSelector),
  getLastTab: () => wrapper.find(tabSelector).last(),
  getNthTab: n => wrapper.find(tabSelector).at(n),
  getNthTabPanel: n => wrapper.find(tabPanelSelector).at(n),
  getNthTabRemove: n => wrapper.find(tabSelector).at(n).find(tabRemoveSelector),
  getAddTab: () => wrapper.find(tabAddSelector),
  clickElement: elem => elem.simulate('click'),
});

describe('Tabs Actions', () => {
  it('tab has activated', () => {
    const wrapper = mount(<App />);
    const page = appSelector(wrapper);
    const lastTab = page.getLastTab();
    page.clickElement(lastTab);
    const newLastTab = page.getLastTab();
    expect(newLastTab).toHaveProp('aria-selected', 'true');
  });

  it('content is visible', () => {
    const wrapper = mount(<App />);
    const page = appSelector(wrapper);
    const tab = page.getNthTab(3);
    page.clickElement(tab);
    const needleContent = page.getNthTabPanel(3);
    expect(needleContent).toContainMatchingElements(1, tabContentSelector);
  });

  it('tab has added', () => {
    const wrapper = mount(<App />);
    const page = appSelector(wrapper);
    const tabs = page.getTabs();
    const lastIndexToAdd = tabs.length;
    const addTab = page.getAddTab();
    page.clickElement(addTab);
    const tabList = page.getTabsList();
    expect(tabList).toContainMatchingElements(lastIndexToAdd + 1, tabSelector);
  });

  it('tab has removed', () => {
    const wrapper = mount(<App />);
    const page = appSelector(wrapper);
    const tabs = page.getTabs();
    const tabsCount = tabs.length;
    const removeTab = page.getNthTabRemove(2);
    page.clickElement(removeTab);
    const tabList = page.getTabsList();
    expect(tabList).toContainMatchingElements(tabsCount - 1, tabSelector);
  });

  it('exact tab has removed', () => {
    const wrapper = mount(<App />);
    const page = appSelector(wrapper);
    const needleTabRemove = page.getNthTabRemove(2);
    page.clickElement(needleTabRemove);
    const needleTab = page.getNthTab(2);
    expect(needleTab).not.toIncludeText('Tab header 3');
  });
});

describe('Work with cookies', () => {
  it('opens last visible tab on reload', () => {
    const TEST_INDEX = 2;
    function CookieContainer() {
      let selectedTab = 0;
      return {
        get() {
          return selectedTab;
        },
        set(i) {
          selectedTab = i;
        },
      };
    }
    const cooks = new CookieContainer();
    Cookies.set.mockImplementation((_, i) => cooks.set(i));
    Cookies.get.mockImplementation(() => cooks.get());

    const wrapper = mount(<App />);
    const page = appSelector(wrapper);
    const needleTab = page.getNthTab(TEST_INDEX);
    page.clickElement(needleTab);

    const wrapper2 = mount(<App />);
    const page2 = appSelector(wrapper2);
    const needleTab2 = page2.getNthTab(TEST_INDEX);
    expect(needleTab2).toHaveProp('aria-selected', 'true');
  });
});
