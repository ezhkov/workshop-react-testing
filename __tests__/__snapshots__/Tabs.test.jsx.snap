// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if tabs renders Check if renders exact tab 1`] = `
<Tabs>
  <Tabs
    defaultFocus={false}
    defaultIndex={null}
    forceRenderTabPanel={false}
    selectedIndex={null}
  >
    <UncontrolledTabs
      className="react-tabs"
      focus={false}
      forceRenderTabPanel={false}
      onSelect={[Function]}
      selectedIndex={1}
    >
      <div
        className="react-tabs"
        data-tabs={true}
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <TabList
          className="react-tabs__tab-list"
          key=".0"
        >
          <ul
            className="react-tabs__tab-list"
            role="tablist"
          >
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-0"
              key=".0"
              panelId="react-tabs-1"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-1"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-0"
                role="tab"
                tabIndex={null}
              >
                Tab header 1
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-2"
              key=".1"
              panelId="react-tabs-3"
              selected={true}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-3"
                aria-disabled="false"
                aria-selected="true"
                className="react-tabs__tab react-tabs__tab--selected"
                id="react-tabs-2"
                role="tab"
                tabIndex="0"
              >
                Tab header 2
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-4"
              key=".2"
              panelId="react-tabs-5"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-5"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-4"
                role="tab"
                tabIndex={null}
              >
                Tab header 3
              </li>
            </Tab>
            <Tab
              className="react-tabs__tab"
              disabledClassName="react-tabs__tab--disabled"
              focus={false}
              id="react-tabs-6"
              key=".3"
              panelId="react-tabs-7"
              selected={false}
              selectedClassName="react-tabs__tab--selected"
              tabRef={[Function]}
            >
              <li
                aria-controls="react-tabs-7"
                aria-disabled="false"
                aria-selected="false"
                className="react-tabs__tab"
                id="react-tabs-6"
                role="tab"
                tabIndex={null}
              >
                Tab header 4
              </li>
            </Tab>
          </ul>
        </TabList>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-1"
          key=".1"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-0"
        >
          <div
            aria-labelledby="react-tabs-0"
            className="react-tabs__tab-panel"
            id="react-tabs-1"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-3"
          key=".2"
          selected={true}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-2"
        >
          <div
            aria-labelledby="react-tabs-2"
            className="react-tabs__tab-panel react-tabs__tab-panel--selected"
            id="react-tabs-3"
            role="tabpanel"
          >
            Tab panel 2 content
          </div>
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-5"
          key=".3"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-4"
        >
          <div
            aria-labelledby="react-tabs-4"
            className="react-tabs__tab-panel"
            id="react-tabs-5"
            role="tabpanel"
          />
        </TabPanel>
        <TabPanel
          className="react-tabs__tab-panel"
          forceRender={false}
          id="react-tabs-7"
          key=".4"
          selected={false}
          selectedClassName="react-tabs__tab-panel--selected"
          tabId="react-tabs-6"
        >
          <div
            aria-labelledby="react-tabs-6"
            className="react-tabs__tab-panel"
            id="react-tabs-7"
            role="tabpanel"
          />
        </TabPanel>
      </div>
    </UncontrolledTabs>
  </Tabs>
</Tabs>
`;

exports[`Check if tabs renders render 1`] = `
<Tabs
  defaultFocus={false}
  defaultIndex={null}
  forceRenderTabPanel={false}
  selectedIndex={null}
>
  <TabList
    className="react-tabs__tab-list"
  >
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Tab header 1
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Tab header 2
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Tab header 3
    </Tab>
    <Tab
      className="react-tabs__tab"
      disabledClassName="react-tabs__tab--disabled"
      focus={false}
      id={null}
      panelId={null}
      selected={false}
      selectedClassName="react-tabs__tab--selected"
    >
      Tab header 4
    </Tab>
  </TabList>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    Tab panel 1 content
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    Tab panel 2 content
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    Tab panel 3 content
  </TabPanel>
  <TabPanel
    className="react-tabs__tab-panel"
    forceRender={false}
    selectedClassName="react-tabs__tab-panel--selected"
  >
    Tab panel 4 content
  </TabPanel>
</Tabs>
`;
